spyder/plugins/ipythonconsole/widgets/spyderapi.py

```diff
@@ -0,0 +1,193 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+Widget that handles communications with a console wanting
+informations from Spyder
+
+Comms transmit data in a list of buffers, and in a json-able dictionnary.
+Here, we only support a buffer list with a single element.
+
+The messages exchanged have the following msg_dict:
+    
+    ```
+    msg_dict = {
+        'spyder_msg_type': spyder_msg_type,
+        'content': content,
+    }
+    ```
+
+The buffer is generated by cloudpickle using `PICKLE_PROTOCOL = 2`.
+
+A protocol to make requests is also implemented:
+The incoming requests have spyder_msg_type = 'blocking_request'.
+The content is:
+
+    ```
+    content = {
+        'request_type': request_type, # The function that should be called
+        'request_args': request_args, # The args of the function
+        'request_kwargs': request_kwargs, # The kwargs of the function
+        'request_id': request_id, # Sent back to link request and reply
+    }
+    ```
+
+The reply has spyder_msg_type = 'blocking_request_reply'.
+The buffer is filled with the data to be sent.
+The content is:
+    
+    ```
+    content = {
+        'request_type': request_type, # The function that has been called
+        'request_id': request_id # The request id
+        'is_error': is_error, # Is the buffer an error to be raised?
+    }
+    ```
+"""
+import cloudpickle
+import logging
+import signal
+
+from qtconsole.rich_jupyter_widget import RichJupyterWidget
+
+from spyder.py3compat import PY2, to_text_string
+
+logger = logging.getLogger(__name__)
+
+# To be able to get and set variables between Python 2 and 3
+PICKLE_PROTOCOL = 2
+
+class ShellApiWidget(RichJupyterWidget):
```

Comment:
What if a third-party plugin wants to access this communication layer? In that case I think this is the wrong abstraction because it shouldn't be a widget.

Instead it could be a normal class or a mixin that can be inherited by the `SpyderPlugin` class we're defining in PR #9226 to expose (at least) `send_kernel_msg`, `handle_kernel_message`, `handle_kernel_request` and `kernel_reply`.

What do you think?

Reply:
The only thing I need in this class that are external are: `kernel_client.comm_manager` and `kernel_manager.signal_kernel`, so changing that should be easy.

`send_kernel_msg` and `kernel_reply` can be exposed, but `handle_kernel_message` and `handle_kernel_request` need to be reimplemented by `NamepaceBrowserWidget` and `ShellWidget`, which are nor plugins. What would be the better way of handling these?